Lombiq.Tests.UI/MonkeyTesting/MonkeyTester.cs

```diff
@@ -0,0 +1,245 @@
+            _options.UrlFilters.All(filter => filter.CanHandle(url, _context));
+
+        private bool TryGetLeftPageToTest(out PageMonkeyTestInfo pageTestInfo)
+        {
+            pageTestInfo = _pageTestInfoList.FirstOrDefault(x => x.HasTimeToTest);
+            return pageTestInfo != null;
+        }
+
+        private PageMonkeyTestInfo GetCurrentPageTestInfo()
+        {
+            var url = _context.Driver.Url;
+            var cleanUrl = CleanUrl(url);
+
+            var pageTestInfo = _pageTestInfoList.FirstOrDefault(x => x.CleanUrl == cleanUrl)
+                ?? new PageMonkeyTestInfo(url, cleanUrl, _options.PageTestTime);
+
+            Log.Info($"Current page is \"{pageTestInfo.CleanUrl}\"");
+
+            return pageTestInfo;
+        }
+
+        private string CleanUrl(string url)
+        {
+            foreach (var cleaner in _options.UrlCleaners)
+                url = cleaner.Handle(url, _context);
+
+            return url;
+        }
+
+        private void TestCurrentPage(PageMonkeyTestInfo pageTestInfo)
+        {
+            int randomSeed = GetRandomSeed();
+
+            TestCurrentPageWithRandomSeed(pageTestInfo, randomSeed);
+        }
+
+        private void TestCurrentPageWithRandomSeed(PageMonkeyTestInfo pageTestInfo, int randomSeed)
+        {
+            try
+            {
+                Log.ExecuteSection(
+                    new LogSection(
+#pragma warning disable S103 // Lines should not be too long
+                       $"Monkey test \"{pageTestInfo.CleanUrl}\" within {pageTestInfo.TimeToTest.ToShortIntervalString()} with {randomSeed} random seed"),
+#pragma warning restore S103 // Lines should not be too long
+                    () =>
+                    {
+                        ExecutePreAssertions();
+
+                        var pageTestTimeLeft = TestCurrentPageAndMeasureTestTimeLeft(pageTestInfo.TimeToTest, randomSeed);
+                        pageTestInfo.TimeToTest = pageTestTimeLeft;
+                        if (!_pageTestInfoList.Contains(pageTestInfo))
+                            _pageTestInfoList.Add(pageTestInfo);
+
+                        ExecutePostAssertions();
+                    });
+            }
+            catch (Exception exception)
+            {
+                throw new AssertionException($"Failure on \"{pageTestInfo.CleanUrl}\" page", exception);
+            }
+        }
+
+        private void ExecutePreAssertions()
+        {
+            if (_options.RunAccessibilityCheckingAssertion) _context.AssertAccessibility();
+
+            if (_options.RunHtmlValidationAssertion) _context.AssertHtmlValidityAsync().GetAwaiter().GetResult();
+        }
+
+        private void ExecutePostAssertions()
+        {
+            if (_options.RunAppLogAssertion) _context.AssertAppLogAsync().GetAwaiter().GetResult();
+
+            if (_options.RunBrowserLogAssertion) _context.AssertBrowserLogAsync().GetAwaiter().GetResult();
```

Comment:
I'm OK regarding inner methods. But `TestCurrentPageAsMonkey()` and `TestCurrentPageAsMonkeyRecursively()` should remain non-async, right? Otherwise it will be uncomfortable to use these methods in tests inside `ExecuteTestAfterSetupAsync`, as `Action<UITestContext> test` parameter is non-async.

Reply:
What I'm saying is that they should be async by default but have sync overloads that make sync calls to them. We mostly need sync code in tests but if someone can use async then let's optimize for that.

Reply:
OK. I've added async method versions.

Reply:
Slowing down is possible by changing the value of `GremlinsAttackDelay` options property, which by default is 10ms.